mapeditor/inspector/towneventswidget.cpp

```diff
@@ -0,0 +1,179 @@
+	}
+	return result;
+}
+
+QVariant toVariant(const CCastleEvent& event)
+{
+	QVariantMap result;
+	result["name"] = QString::fromStdString(event.name);
+	result["message"] = QString::fromStdString(event.message.toString());
+	result["players"] = QVariant::fromValue(event.players);
+	result["humanAffected"] = QVariant::fromValue(event.humanAffected);
+	result["computerAffected"] = QVariant::fromValue(event.computerAffected);
+	result["firstOccurrence"] = QVariant::fromValue(event.firstOccurrence);
+	result["nextOccurrence"] = QVariant::fromValue(event.nextOccurrence);
+	result["resources"] = toVariant(event.resources);
+	result["buildings"] = toVariant(event.buildings);
+	result["creatures"] = toVariant(event.creatures);
+
+	return QVariant(result);
+}
+
+CCastleEvent eventFromVariant(CMapHeader& map, CGTownInstance& town, const QVariant& variant)
+{
+	CCastleEvent result;
+	auto v = variant.toMap();
+	result.name = v.value("name").toString().toStdString();
+	result.message.appendTextID(mapRegisterLocalizedString("map", map, TextIdentifier("town", town.instanceName, "event", result.name, "message"), v.value("message").toString().toStdString()));
+	result.players = v.value("players").toInt();
+	result.humanAffected = v.value("humanAffected").toInt();
+	result.computerAffected = v.value("computerAffected").toInt();
+	result.firstOccurrence = v.value("firstOccurrence").toInt();
+	result.nextOccurrence = v.value("nextOccurrence").toInt();
+	result.resources = resourcesFromVariant(v.value("resources"));
+	result.buildings = buildingsFromVariant(v.value("buildings"));
+	result.creatures = creaturesFromVariant(v.value("creatures"));
+	return result;
+}
+
+void TownEventsWidget::obtainData()
+{
+	for (const auto & event : town.events)
+	{
+		auto eventName = QString::fromStdString(event.name);
+		auto itemText = QString::fromStdString("Day %1 - %2").arg(event.firstOccurrence+1, 3).arg(eventName);
+
+		auto * item = new QListWidgetItem(itemText);
+		item->setData(Qt::UserRole, toVariant(event));
+		ui->eventsList->addItem(item);
+	}
+}
+
+void TownEventsWidget::commitChanges(MapController& controller)
+{
+	town.events.clear();
+	for (int i = 0; i < ui->eventsList->count(); ++i)
+	{
+		const auto * item = ui->eventsList->item(i);
+		town.events.push_back(eventFromVariant(*controller.map(), town, item->data(Qt::UserRole)));
+	}
+}
+
+void TownEventsWidget::on_timedEventAdd_clicked()
+{
+	CCastleEvent event;
+	event.name = tr("New event").toStdString();
+	auto* item = new QListWidgetItem(QString::fromStdString(event.name));
+	item->setData(Qt::UserRole, toVariant(event));
+	ui->eventsList->addItem(item);
+	on_eventsList_itemActivated(item);
+}
+
+void TownEventsWidget::on_timedEventRemove_clicked()
+{
+	if (auto* item = ui->eventsList->currentItem())
+		ui->eventsList->takeItem(ui->eventsList->row(item));
+}
+
+void TownEventsWidget::on_eventsList_itemActivated(QListWidgetItem* item)
+{
+	new TownEvent(town, item, parentWidget());
```

Comment:
is it supposed to show modal dialog?

Reply:
Yes

Reply:
then the right way is:
```cpp
TownEventDialog dlg{town, item, parentWidget()};
dlg.exec();
```

and drop that `show()` from constructor